metadata-ingestion-modules/airflow-plugin/src/datahub_airflow_plugin/datahub_listener.py

```diff
@@ -0,0 +1,411 @@
+    # Using globals instead of functools.lru_cache to make testing easier.
+    global _airflow_listener_initialized
+    global _airflow_listener
+
+    if not _airflow_listener_initialized:
+        _airflow_listener_initialized = True
+
+        plugin_config = get_lineage_config()
+
+        if plugin_config.enabled:
+            _airflow_listener = DataHubListener(config=plugin_config)
+
+            if plugin_config.disable_openlineage_plugin:
+                # Deactivate the OpenLineagePlugin listener to avoid conflicts.
+                from openlineage.airflow.plugin import OpenLineagePlugin
+
+                OpenLineagePlugin.listeners = []
+
+    return _airflow_listener
+
+
+def run_in_thread(f):
+    @functools.wraps(f)
+    def wrapper(*args, **kwargs):
+        try:
+            if _RUN_IN_THREAD:
+                thread = threading.Thread(
+                    target=f, args=args, kwargs=kwargs, daemon=True
+                )
+                thread.start()
+
+                thread.join(timeout=10)
+            else:
+                f(*args, **kwargs)
+        except Exception as e:
+            logger.exception(e)
+
+    return wrapper
+
+
+class DataHubListener:
+    __name__ = "DataHubListener"
+
+    def __init__(self, config: DatahubLineageConfig):
+        self.config = config
+        self._set_log_level()
+
+        self._emitter = config.make_emitter_hook().make_emitter()
+        self._graph = None
+        logger.info(f"DataHub plugin using {repr(self._emitter)}")
+
+        # See discussion here https://github.com/OpenLineage/OpenLineage/pull/508 for
+        # why we need to keep track of tasks ourselves.
+        self._task_holder = TaskHolder()
+
+        # In our case, we also want to cache the initial datajob object
+        # so that we can add to it when the task completes.
+        self._datajob_holder: Dict[str, DataJob] = {}
+
+        self.extractor_manager = ExtractorManager()
+
+    @property
+    def emitter(self):
+        return self._emitter
+
+    @property
+    def graph(self) -> Optional["DataHubGraph"]:
+        if self._graph:
+            return self._graph
+
+        if isinstance(self._emitter, DatahubRestEmitter):
+            # This is lazy initialized to avoid throwing errors on plugin load.
+            self._graph = self._emitter.to_graph()
+            self._emitter = self._graph
+
+    def _set_log_level(self) -> None:
+        if self.config.log_level:
+            logging.getLogger(__name__.split(".")[0]).setLevel(self.config.log_level)
+        if self.config.debug_emitter:
+            logging.getLogger("datahub.emitter").setLevel(self.config.log_level)
+
+    def _make_emit_callback(self) -> Callable[[Optional[Exception], str], None]:
+        def emit_callback(err: Optional[Exception], msg: str) -> None:
+            if err:
+                logger.error(f"Error sending metadata to datahub: {msg}", exc_info=err)
```

Comment:
Since this doesn't use `self`, can this just be a static or standalone method on its own?